app/uk/gov/hmrc/exports/metrics/ExportsMetrics.scala

```diff
@@ -18,23 +18,133 @@ package uk.gov.hmrc.exports.metrics
+    val result = block
+    timerContext.stop()
+    result
+  }
+
+  def timeAsyncCall[A](timer: Timer)(block: => Future[A])(implicit ec: ExecutionContext): Future[A] = {
+    val timerContext: Context = startTimer(timer)
+    val timerRunning: AtomicBoolean = new AtomicBoolean(true)
+
+    try {
+      val result = block
+
+      result.foreach { _ =>
+        if (timerRunning.compareAndSet(true, false)) {
+          timerContext.stop()
+        }
+      }
+
+      result.failed.foreach { _ =>
+        if (timerRunning.compareAndSet(true, false)) {
+          timerContext.stop()
+        }
+      }
+
+      result
+    } catch {
+      case NonFatal(e) =>
+        if (timerRunning.compareAndSet(true, false)) {
+          timerContext.stop()
+        }
+        throw e
+    }
+  }
+
+  def timeAsyncCall[A](monitor: Monitor)(block: => Future[A])(implicit ec: ExecutionContext): Future[A] = {
+    incrementCounter(monitor.callCounter)
+    val timerContext: Context = startTimer(monitor.timer)
+    val timerRunning: AtomicBoolean = new AtomicBoolean(true)
 
-  val counters = Map(notificationMetric -> metrics.defaultRegistry.counter(s"$notificationMetric.counter"))
+    try {
+      val result = block
 
-  def startTimer(feature: String): Context = timers(feature).time()
+      result.foreach { _ =>
+        if (timerRunning.compareAndSet(true, false)) {
+          timerContext.stop()
+          incrementCounter(monitor.completionCounter)
+        }
+      }
 
-  def incrementCounter(feature: String): Unit = counters(feature).inc()
+      result.failed.foreach { _ =>
+        if (timerRunning.compareAndSet(true, false)) {
+          timerContext.stop()
+          incrementCounter(monitor.failureCounter)
+        }
+      }
+
+      result
+    } catch {
+      case NonFatal(e) =>
+        if (timerRunning.compareAndSet(true, false)) {
+          timerContext.stop()
+          incrementCounter(monitor.failureCounter)
+        }
+        throw e
+    }
+  }
 
 }
 
-object MetricIdentifiers {
-  val notificationMetric = "submission.notification"
+object ExportsMetrics {
+
+  case class Meter(name: String) { val path = s"$name.rate" }
+  case class Timer(name: String) { val path = s"$name.timer" }
+  case class Counter(name: String) { val path = s"$name.counter" }
+
+  trait Monitor {
+    val name: String
+
+    val path: String = s"$name.monitor"
+    val timer: Timer = Timer(path)
+    val callCounter: Counter = Counter(path)
+    val completionCounter: Counter = Counter(s"$path.success")
+    val failureCounter: Counter = Counter(s"$path.failed")
+  }
+
+  object Timers {
+    val notificationTimer: Timer = Timer(notificationMetric)
+
+    val upstreamCustomsDeclarationsTimer: Timer = Timer(upstreamCustomsDeclarationsMetric)
+
+    val declarationFindAllTimer: Timer = Timer(declarationFindAllMetric)
+    val declarationFindSingleTimer: Timer = Timer(declarationFindSingleMetric)
+    val declarationUpdateTimer: Timer = Timer(declarationUpdateMetric)
```

Comment:
To aid readability I'd suggest changing these from e.g. 'declarationFindAllMetric' (that sounds like 'find all metrics') to 'metricsFindAllDeclarations', 'metricsFindSingleDeclaration' & 'metricsUpdateDeclarations'

Reply:
I've spent quite a bit of time figuring out what naming convention to use. My thinking is that metrics relating to similar processes have their prefixes the same. Having 'metric' at the end sounds more natural to me, although I am aware English is not my first language ;) 

One solution I am considering is to create a structure of nested objects, similar to how files are organised in folders. I'll give it a try and see how it looks like.